src/auth/auth.js

```diff
@@ -0,0 +1,66 @@
+import express from 'express';
+import passport from 'passport';
+import * as oauth2 from 'passport-oauth2';
+
+export default function authentication(config) {
+  const app = express();
+
+  passport.use(new oauth2.Strategy({
+    authorizationURL: config.oauthAuthorizationURL,
+    tokenURL: config.oauthTokenURL,
+    clientID: config.oauthClientID,
+    clientSecret: config.oauthClientSecret,
+    callbackURL: config.serverRootURL + '/auth/login/callback'
+  },
+    (accessToken, refreshToken, profile, cb) => {
+      return cb(null, {accessToken, refreshToken, profile});
+    }
+  ));
+
+  app.use(passport.initialize());
+  app.use(passport.session());
+
+  passport.serializeUser((user, cb) => {
+    cb(null, user.accessToken);
+  });
+
+  passport.deserializeUser((accessToken, cb) => {
+    cb(null, {accessToken});
+  });
+
+  app.get('/auth/login', passport.authenticate('oauth2', {
+    scope: [
+      'cloud_controller.read',
+      'openid',
+      'uaa.user'
+    ]
+  }));
+
+  app.get('/auth/login/callback', passport.authenticate('oauth2'), (req, res) => {
+    const redirection = req.session.returnTo;
+    delete req.session.returnTo;
+
+    res.redirect(redirection || '/');
+    req.log.info({
+      returnTo: redirection
+    }, `Authentication successful, redirecting back to ${redirection}`);
+  });
+
+  app.get('/auth/logout', (req, res) => {
+    req.logout();
+    res.redirect('/');
+  });
+
+  app.use((req, res, next) => {
+    if (req.isAuthenticated()) {
```

Comment:
low priority: 

maybe we could set `req.accessToken = req.session.passport.user` here as a shortcut to the rather verbose `req.session.passport.user` for consumers of the auth middleware.

It's also a bit nicer to be able to do:

```
if (req.accessToken) { ... }
```

than 

```
if(req.session && req.session.passport && req.session.passport.user) { ... }
```